src/server/util/authenticationHelpers.js

```diff
@@ -1,24 +1,79 @@
 const { Pool } = require('pg');
 const { sql } = require('./sqlHelpers');
-const { saltHashPassword } = require('./pwHash');
+const { saltHashPassword, validateHashPassword } = require('./pwHash');
 
 const pool = new Pool();
 
-const getUser = async function(req, res, next) {
-    const { username, email } = req.body;
+const getUserInDatabase = async function(username, email) {
     const client = await pool.connect();
-
     const result = await client.query(sql`SELECT * FROM users WHERE email = ${email} OR username = ${username};`);
     client.release();
+    return result.rows;
+}
+
+const userExistsInDatabase = async function(username, email) {
+    const rows = await getUserInDatabase(username, email);
+    let user, exists = rows.length > 0;
+    if (exists) {
+        user = rows[0];
+        if (username) exists = exists && user.username === username;
+        if (email) exists = exists && user.email === email;
+    }
+        
+    return { exists, user };
+}
 
-    if(result.rows.length > 0) {
+const duplicatedUser = async function(req, res, next) {
+    const { username, email } = req.body;
+    const { exists } = await userExistsInDatabase(username, email);
+    if(exists) {
         res.status(409).send({ error: 'Duplicated user. User already exists!' });
         return;
     }
 
     next();
 }
 
+const loginUser = async function(req, res, next) {
+    const { username, email, password } = req.body;
+    const { exists, user } = await userExistsInDatabase(username, email);
+
+    // Case when user is found
+    if (exists) {
+        const { salt, hash } = user;
+        
+        // Case of anonymous user, where only the username / email stored
+        if (!password && !hash && !salt) {
+            req.anonymous = true;
+            next();
+            return;
```

Comment:
I don't think we need a `return` statement after `next()`

Reply:
you definitely do

Reply:
@ktbee without `return`, the rest of the code after `next()` will run eventually

Reply:
Oh interesting, I thought express moved on after next(). Thanks for letting me know!